tutorials/3.Interactions.md

```diff
@@ -0,0 +1,348 @@
+
+We're going to change up how `myPing` is formed, so we can see how to make both a slash command and a text command.
+
+First, we separate out the function that responds with `pong`, and make it so that instead of instantly sending the message, it generates a `MessageDetails` data structure that represents such a message.
+
+```haskell
+myPing :: Command
+myPing = Command "myping" (parseComm myPingAction) []
+
+myPingAction :: DatabaseDiscord MessageDetails 
+myPingAction = return $ messageDetailsBasic "pong"
+```
+
+Due to some of the setup that's been done separately, this will work as before, replying with "pong" when we call "myping".
+
+We can now create our first application command!
+
+To do this, we have to tell Discord what application command to create, and then we have to come up with a way to answer that. Luckily, we've made some magic to make some of this somewhat easier, and package it all up for us.
+
+```haskell
+myPingInteraction :: Maybe ApplicationCommandRecv
+myPingInteraction = makeApplicationCommandPair "myping" "my unique ping" myPingAction
+```
+
+Here we've (possibly) created an application command called "myping", with a description of "my unique ping" that is created from `myPingAction`.
+
+If we add `myPingInteraction` to the `applicationCommands` list in `pingPlugin'`, and run our bot, we can see that we can now (after a short delay) type `/myping` into discord, press enter, and get a special response in return.
+
+Well, that's pretty exciting. But we could do that before. What do we need to do to get the rest of our ping functions up to scratch?
+
+Turns out there's a fair amount of information that we took for granted before that we're now going to have to work through, but don't worry, we're gonna get through it step by step.
+
+### Labelled arguments
+
+Slash commands unfortunately need to have each of their parameters named, which means that anything we give to our function has to be named.
+
+Let's uncomment `myEcho` and roll up our sleeves to see what we want to do here.
+
+First we pull out `echoHelp` into its own top level function, modify the return and arguments so it returns a message that is just the text we pass in - and we run into an issue. Having a `RestOfInput` doesn't make any sense in a slash command, so we can't use `echoHelp` as is. We're going to have to construct two subtly different functions that we can process differently to get what we want.
+
+```haskell
+-- current `echoHelp`
+echoHelp :: RestOfInput Text -> DatabaseDiscord MessageDetails
+echoHelp (ROI t) = return $ messageDetailsBasic t
+```
+
+We take away the restriction of `RestOfInput` for now, breaking `myecho`, but we'll fix it shortly. 
+
+We then change `myEcho` so that instead of having `(parseComm echoHelp)`, we instead have `(parseComm (echoHelp . unROI))`. This function is now the same as it originally was, and we can reuse our new `echoHelp` for our interactions.
```

Comment:
It would be nice to have a tiny mention of what `unROI` is and why it is necessary